src/lib/shmem/src/shmalloc_impl.rs

```diff
@@ -0,0 +1,739 @@
+//! allocation, but should be OK when used for just a few types.
+//!
+//! This code is intended to be private; the `allocator` module is the public, safer-to-use
+//! front end.
+
+use crate::raw_syscall::*;
+use numtoa::NumToA;
+
+const PATH_MAX_NBYTES: usize = 255;
+pub(crate) const PATH_BUF_NBYTES: usize = PATH_MAX_NBYTES + 1;
+pub(crate) type PathBuf = [u8; PATH_BUF_NBYTES];
+
+use vasi::VirtualAddressSpaceIndependent;
+
+fn get_null_path_buf() -> PathBuf {
+    [0; PATH_BUF_NBYTES]
+}
+
+#[cfg(debug_assertions)]
+type CanaryBuf = [u8; 4];
+
+#[cfg(debug_assertions)]
+fn get_canary_buf() -> CanaryBuf {
+    [0xDE, 0xAD, 0xBE, 0xEF]
+}
+
+#[cfg(not(debug_assertions))]
+type CanaryBuf = [u8; 0];
+
+#[cfg(not(debug_assertions))]
+fn get_canary_buf() -> CanaryBuf {
+    []
+}
+
+trait Canary {
+    fn canary_init(&mut self);
+    fn canary_check(&self) -> bool;
+
+    #[cfg(debug_assertions)]
+    fn canary_assert(&self) {
+        assert!(self.canary_check());
+    }
+
+    #[cfg(not(debug_assertions))]
+    fn canary_assert(&self) {}
+}
+
+pub(crate) enum AllocError {
+    Clock,
+    Open,
+    FTruncate,
+    MMap,
+    MUnmap,
+    Unlink,
+    WrongAllocator,
+    Leak,
+}
+
+pub(crate) fn log_error(error: AllocError, errno: Option<i32>) {
+    const STDERR_FD: i32 = 2;
+    let mut errno_buf = [0u8; 32];
+
+    let err: &[u8] = match error {
+        AllocError::Clock => b"[shadow:shmalloc] [CRITICAL] Error calling clock_gettime().\n",
+        AllocError::Open => b"[shadow:shmalloc] [CRITICAL] Error calling open().\n",
+        AllocError::FTruncate => b"[shadow:shmalloc] [CRITICAL] Error calling ftruncate().\n",
+        AllocError::MMap => b"[shadow:shmalloc] [CRITICAL] Error calling mmap().\n",
+        AllocError::MUnmap => b"[shadow:shmalloc] [WARNING] Error calling munmap().\n",
+        AllocError::Unlink => b"[shadow:shmalloc] [WARNING] Error calling unlink().\n",
+        AllocError::WrongAllocator => b"[shadow:shmalloc] [ERROR] Block was passed to incorrect allocator.\n",
+        AllocError::Leak => b"[shadow:shmalloc] [ERROR] Allocator destroyed but not all blocks are deallocated first.\n",
+    };
+
+    unsafe {
+        write(STDERR_FD, err.as_ptr() as *const _, err.len());
+    }
+
+    if let Some(e) = errno {
+        let errno_buf = e.numtoa(10, &mut errno_buf);
+
+        const PREAMBLE: &[u8; 44] = b"[shadow:shmalloc] [WARNING] Last error num: ";
+
+        unsafe {
+            write(STDERR_FD, PREAMBLE.as_ptr() as *const _, PREAMBLE.len());
+        }
+        unsafe {
+            write(STDERR_FD, errno_buf.as_ptr() as *const _, errno_buf.len());
+        }
+        unsafe {
+            write(STDERR_FD, b"\n".as_ptr() as *const _, 1);
+        }
+    }
+}
+
+pub(crate) fn log_error_and_exit(error: AllocError, errno: Option<i32>) -> ! {
+    log_error(error, errno);
+    let _ = kill(0, linux_api::signal::Signal::SIGABRT.into());
```

Comment:
`kill` with pid=0 doesn't do what you expect. This is a mistake I've made before, with "fun" results :)

kill(2):

> If pid equals 0, then sig is sent to every process in the process group of the calling process.

Also we're not guaranteed that the signal will be delivered to the current thread, in which case `kill` might return before the signal is actually delivered.

I think what we actually want is something like: `tgkill(getpid(), gettid(), SIGABRT)`. https://shadow.github.io/docs/rust/linux_api/signal/fn.tgkill.html

Reply:
fixed